File: samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/zone/handlers/BorderSubscriptionHandlers.java
package systems.zlink.samples.zoneworld.dynamic;

import systems.zlink.framework.handlers.ZLinkSpotSubscription;
import systems.zlink.framework.spots.ZLinkSpotSubscriptionHandler;
import systems.zlink.samples.zoneworld.server.zone.spots.ZoneSpot;
import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class BorderSubscriptionHandlers {
    private BorderSubscriptionHandlers() {}

    // --8<-- [start:doc-zw-border-subscribe]
    public static Class<?> forRoute(String fromZoneId, String toZoneId) {
        return switch (ZoneWorldNames.borderTopic(fromZoneId, toZoneId)) {
            case ZoneWorldNames.NW_NE -> NorthWestToNorthEast.class;
            case ZoneWorldNames.NW_SW -> NorthWestToSouthWest.class;
            case ZoneWorldNames.NE_NW -> NorthEastToNorthWest.class;
            case ZoneWorldNames.NE_SE -> NorthEastToSouthEast.class;
            case ZoneWorldNames.SW_NW -> SouthWestToNorthWest.class;
            case ZoneWorldNames.SW_SE -> SouthWestToSouthEast.class;
            case ZoneWorldNames.SE_NE -> SouthEastToNorthEast.class;
            case ZoneWorldNames.SE_SW -> SouthEastToSouthWest.class;
            default ->
                    throw new IllegalArgumentException(
                            "unknown border route: " + fromZoneId + " -> " + toZoneId);
        };
    }

    // --8<-- [end:doc-zw-border-subscribe]

    private static CompletionStage<Void> apply(ZoneSpot spot, Messages.ZoneBorderEvent event) {
        spot.applyBorder(event);
        return CompletableFuture.completedFuture(null);
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_NE)
    public static final class NorthWestToNorthEast
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_SW)
    public static final class NorthWestToSouthWest
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_NW)
    public static final class NorthEastToNorthWest
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_SE)
    public static final class NorthEastToSouthEast
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_NW)
    public static final class SouthWestToNorthWest
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_SE)
    public static final class SouthWestToSouthEast
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_NE)
    public static final class SouthEastToNorthEast
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_SW)
    public static final class SouthEastToSouthWest
            implements ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        @Override
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) {
            return apply(spot, event);
        }
    }
}

File: samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/dynamic/BorderSubscriptionHandlers.kt
package systems.zlink.samples.kotlin.zoneworld.dynamic

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.handlers.ZLinkSpotSubscription
import systems.zlink.framework.spots.ZLinkSpotSubscriptionHandler
import systems.zlink.samples.kotlin.zoneworld.server.zone.ZoneSpot
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

class BorderSubscriptionHandlers {
    companion object {
        fun forRoute(from: String, to: String): Class<*> =
            when (ZoneWorldNames.borderTopic(from, to)) {
                ZoneWorldNames.NW_NE -> NorthWestToNorthEast::class.java
                ZoneWorldNames.NW_SW -> NorthWestToSouthWest::class.java
                ZoneWorldNames.NE_NW -> NorthEastToNorthWest::class.java
                ZoneWorldNames.NE_SE -> NorthEastToSouthEast::class.java
                ZoneWorldNames.SW_NW -> SouthWestToNorthWest::class.java
                ZoneWorldNames.SW_SE -> SouthWestToSouthEast::class.java
                ZoneWorldNames.SE_NE -> SouthEastToNorthEast::class.java
                ZoneWorldNames.SE_SW -> SouthEastToSouthWest::class.java
                else -> error("unknown border route")
            }

        fun apply(spot: ZoneSpot, event: Messages.ZoneBorderEvent): CompletionStage<Void> {
            spot.applyBorder(event)
            return CompletableFuture.completedFuture(null)
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_NE)
    class NorthWestToNorthEast : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_SW)
    class NorthWestToSouthWest : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_NW)
    class NorthEastToNorthWest : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_SE)
    class NorthEastToSouthEast : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_NW)
    class SouthWestToNorthWest : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_SE)
    class SouthWestToSouthEast : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_NE)
    class SouthEastToNorthEast : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_SW)
    class SouthEastToSouthWest : ZLinkSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) = apply(spot, event)
    }
}
